### PR TITLE
[clang] Use internal linkage for c23 constexpr vars.

### LINKED ISSUE #97830: C23 constexpr should not have external linkage
Hi,
thanks for implementing `constexpr` for clang-19!

The following code

```{.C}
constexpr int category_width = 5;
int const w = 5;
int const* a = &category_width;
```
compiles, but produces an external symbol for `category_width` and then multiple defined symbol errors on linkage. So I cannot use `constexpr` in header files, or I'd have to prefix it with `static`, which defeats the purpose of `constexpr` ...

C23 has in 6.2.2 p3

> If the declaration of a file scope identifier for:
> — an object contains any of the storage-class specifiers `static` or `constexpr`;
> — or, a function contains the storage-class specifier `static`,
> then the identifier has internal linkage.

Thanks
Jens

### PR DESCRIPTION
**Pull request by chestnykh**
See C23 std 6.2.2p3.
Fix #97830

### CODE REVIEW

File: clang/lib/Parse/ParseDecl.cpp [hunk 1]
```diff
@@ -4428,10 +4428,14 @@ void Parser::ParseDeclarationSpecifiers(
...
     case tok::kw_constexpr:
-      if (getLangOpts().C23)
+      if (getLangOpts().C23) {
         Diag(Tok, diag::warn_c23_compat_keyword) << Tok.getName();
-      isInvalid = DS.SetConstexprSpec(ConstexprSpecKind::Constexpr, Loc,
-                                      PrevSpec, DiagID);
+        isInvalid = DS.SetConstexprSpec(ConstexprSpecKind::Constexpr, Loc,
+                                        PrevSpec, DiagID);
+
+        isInvalid = DS.SetStorageClassSpec(Actions, DeclSpec::SCS_static, Loc,
```
Comment (tschuett) [inline clang/lib/Parse/ParseDecl.cpp:4436]: Are you sure that you want to assign `isInvalid` twice?

Comment (chestnykh) [reply clang/lib/Parse/ParseDecl.cpp:4436]: I need to rework this PR a little bit

Comment (chestnykh) [reply clang/lib/Parse/ParseDecl.cpp]: Done

**Review by Fznamznon — APPROVED** [review]
LGTM, though please let others to take a look too.

**Review by AaronBallman — APPROVED** [review]
LGTM but please add a release note to clang/docs/ReleaseNotes.rst so users know about the fix as well.

Once that's added, will you need someone to land the changes on your behalf?

**Comment by chestnykh** [comment]
> LGTM but please add a release note to clang/docs/ReleaseNotes.rst so users know about the fix as well.
> 
> Once that's added, will you need someone to land the changes on your behalf?

@AaronBallman , yes, i will need this. I'm going to add release notes right now.

**Comment by Fznamznon** [comment]
> LGTM but please add a release note to clang/docs/ReleaseNotes.rst so users know about the fix as well.

Note that only clang 19 has `constexpr` support in C23. https://godbolt.org/z/4bThdsqK5

**Comment by chestnykh** [comment]
Release notes added

**Comment by AaronBallman** [comment]
> > LGTM but please add a release note to clang/docs/ReleaseNotes.rst so users know about the fix as well.
> 
> Note that only clang 19 has `constexpr` support in C23. https://godbolt.org/z/4bThdsqK5 Release note saying that support has been implemented is already there.

Ooh oops, I thought we had this in Clang 18

**Comment by chestnykh** [comment]
> > > LGTM but please add a release note to clang/docs/ReleaseNotes.rst so users know about the fix as well.
> > 
> > 
> > Note that only clang 19 has `constexpr` support in C23. https://godbolt.org/z/4bThdsqK5 Release note saying that support has been implemented is already there.
> 
> Ooh oops, I thought we had this in Clang 18

Should I change anything?

**Comment by AaronBallman** [comment]
> > > > LGTM but please add a release note to clang/docs/ReleaseNotes.rst so users know about the fix as well.
> > > 
> > > 
> > > Note that only clang 19 has `constexpr` support in C23. https://godbolt.org/z/4bThdsqK5 Release note saying that support has been implemented is already there.
> > 
> > 
> > Ooh oops, I thought we had this in Clang 18
> 
> Should I change anything?

Sorry for the back-and-forth -- no new release note needed, so you can remove what you added.

**Comment by chestnykh** [comment]
> > > > > LGTM but please add a release note to clang/docs/ReleaseNotes.rst so users know about the fix as well.
> > > > 
> > > > 
> > > > Note that only clang 19 has `constexpr` support in C23. https://godbolt.org/z/4bThdsqK5 Release note saying that support has been implemented is already there.
> > > 
> > > 
> > > Ooh oops, I thought we had this in Clang 18
> > 
> > 
> > Should I change anything?
> 
> Sorry for the back-and-forth -- no new release note needed, so you can remove what you added.

Removed :)

**Pull request merged**